resources/prosody-plugins/mod_av_moderation_component.lua

```diff
@@ -0,0 +1,242 @@
+                else
+                    room.av_moderation = {};
+                    room.av_moderation_actors = {};
+                    room.av_moderation[mediaType] = {};
+                    room.av_moderation_actors[mediaType] = occupant.nick;
+                end
+            else
+                enabled = false;
+                if not room.av_moderation and not room.av_moderation[mediaType] then
+                    module:log('warn', 'Concurrent moderator enable/disable request or something is out of sync');
+                else
+                    room.av_moderation[mediaType] = nil;
+                    room.av_moderation_actors[mediaType] = nil;
+                end
+            end
+
+            -- send message to all occupants
+            notify_occupants_enable(nil, enabled, room, occupant.nick, mediaType);
+            return true;
+        elseif moderation_command.attr.jidToWhitelist and room.av_moderation then
+            local occupant_jid = moderation_command.attr.jidToWhitelist;
+            -- check if jid is in the room, if so add it to whitelist
+            -- inform all moderators and admins and the jid
+            local occupant = room:get_occupant_by_nick(occupant_jid);
+
+            if not occupant then
+                module:log('warn', 'No occupant %s found for %s', occupant_jid, room.jid);
+                return false;
+            end
+
+            local whitelist = room.av_moderation[mediaType];
+            if not whitelist then
+                whitelist = {};
+                room.av_moderation[mediaType] = whitelist;
+            end
+            table.insert(whitelist, occupant_jid);
+
+            notify_whitelist_change(occupant.jid, true, room, mediaType);
+
+            return true;
+        end
+    end
+
+    -- return error
+    return false
+end
+
+-- handles new occupants to inform them about the state enabled/disabled, new moderators also get and the whitelist
+function occupant_joined(event)
+    local room, occupant = event.room, event.occupant;
+
+    if is_healthcheck_room(room.jid) then
+        return;
+    end
+
+    if room.av_moderation then
+        for _,mediaType in pairs({'audio', 'video'}) do
+            if room.av_moderation[mediaType] then
+                notify_occupants_enable(
+                    occupant.jid, true, room, room.av_moderation_actors[mediaType], mediaType);
+            end
+        end
+
+        -- if moderator send the whitelist
+        if occupant.role == 'moderator' then
+            notify_whitelist_change(occupant.jid, false, room);
+        end
+    end
+end
+
+-- when a occupant was granted moderator we need to update him with the whitelist
+function occupant_affiliation_changed(event)
+    -- if set from someone and is owner (and that is not jicofo - not admin)
+    if event.actor and event.affiliation == 'owner' and not is_admin(event.actor) then
```

Comment:
Does this actually check that a member was granted moderator? And why does it matter who the actor was?

Reply:
As this handles the case when someone does "grant moderator" in the UI.
Hum, but I haven't tested the case where someone becomes a moderator, as one left and jicofo grants it.
So maybe this check is not needed ... will test it while testing the rest of the changes.

Reply:
Fixed it, I think there is a bug in the way we use set_affiliation in allowners and the data in the event sent to occupant-joined is not reflecting the changes. I added a note in the code in the relevant place.

This code is aiming to be triggered only when someone grants moderator rights.